### ISSUE TITLE
Debug Viewer 

### ISSUE THREAD

**Issue by lukasmartinelli** [geometalab/osm2vectortiles#12]
Create a debug viewer with MapBoxGL JS.

It downloads tile json of particular tileset.
Prints the list of layers, on each layer you apply a color of a palette.
List of colors are applied on different layers as different colors.
Simulate with Mapbox GL JS what you in Debug Mode inside the new Mapbox Studio:

<img width="795" alt="screen shot 2015-09-07 at 13 39 21" src="https://cloud.githubusercontent.com/assets/59284/9715996/0599710c-5566-11e5-95e8-2d6bd7e9bce5.png">

On Mapbox GL JS has an example how to select features:
https://www.mapbox.com/mapbox-gl-js/example/featuresat/

If you host vector tiles you need to show something - which is the debug viewer.
Give the debugger a link to tiles.json and then it displays the debug view.

**Comment by lukasmartinelli** [comment]
Perhaps this is just a cake walk with tessera.
https://github.com/mojodna/tilelive-xray

**Comment by klokan** [comment]
Another tool which may be helpful for us: http://lxbarth.com/compare/ & https://github.com/mapbox/compare

**Comment by klokan** [comment]
@petrsloup this is the discussed x-ray debug viewer.
The default colours are derived from names of layers: see https://github.com/mapbox/tilelive-vector/pull/51/files#diff-168726dbe96b3ce427e7fedce31bb0bcR536

**Comment by klokan** [comment]
The first version of the online debug-viewer is now implemented in OpenLayers3 - thanks @petrsloup

There are two tools, both accepting after a hash the url of a vector tileset TileJSON:

- **x-ray preview** with possibility to turn on/off individual layers and with mouse hover study the values of features under the cursor.
See:
http://klokantech.github.io/ol3-sandbox/vector/xray.html#http://tileserver.maptiler.com/zurich.json

- **tile inspector**: which allows to interactively pick a tile and extract relevant information such as the size in bytes, list of layers, number of features, classes, types, etc.
See:
http://klokantech.github.io/ol3-sandbox/vector/tile-inspector.html#http://tileserver.maptiler.com/zurich.json

On both tools, it is good to turn off the heaviest layers before you interact with the map, because most of the parsing/displaying is made on (blocking) CPU in OpenLayers without the hardware acceleration of MapBox GL JS - and our tiles are sometimes huge!

Try it and comment.

**Comment by lukasmartinelli** [comment]
This is amazing. Huge thanks to @petrsloup.

Especially the tile inspector will be a big help at this stage. We will use this to take closer look at the Mapbox layers too see which layers are visible where and how many are shown (a visual approach to our tile comparing).

The X-Ray preview has similar features to Mapbox Studio Debug Viewer but more practical to use because it shows the values on `hover` and we don't have to always try to click on tiny elements.

Is the source also available so we can look at it / contribute?
If not it is not that bad because we can just modify the source of the page and point it to Mapbox Streets.

**Comment by klokan** [comment]
Source code is of course available, feel free to fork, patch, make pull
request:

https://github.com/klokantech/ol3-sandbox

On Wednesday, 4 November 2015, Lukas Martinelli <notifications@github.com>
wrote:

> This is amazing. Huge thanks to @petrsloup <https://github.com/petrsloup>.
>
> Especially the tile inspector will be a big help at this stage. We will
> use this to take closer look at the Mapbox layers too see which layers are
> visible where and how many are shown (a visual approach to our tile
> comparing).
>
> The X-Ray preview has similar features to Mapbox Studio Debug Viewer but
> more practical to use because it shows the values on hover and we don't
> have to always try to click on tiny elements.
>
> Is the source also available so we can look at it / contribute?
> If not it is not that bad because we can just modify the source of the
> page and point it to Mapbox Streets.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/geometalab/osm2vectortiles/issues/12#issuecomment-153595605>
> .
>


-- 
Petr Pridal, Ph.D.
CEO

Klokan Technologies GmbH
Hofnerstrasse 98, 6314 Unterageri, Switzerland
Tel: +41 (0)41 511 26 12
Email: info@klokantech.com
Web: http://www.klokantech.com/

**Comment by sfkeller** [comment]
Awesome work, guys!

**Comment by lukasmartinelli** [comment]
Just the links. Great trick to allow appending the TileJSON url to the url hashbang btw.

Mapbox Streets X-Ray Preview:

http://klokantech.github.io/ol3-sandbox/vector/xray.html#http://a.tiles.mapbox.com/v4/mapbox.mapbox-streets-v5.json?access_token=pk.eyJ1Ijoib3NtMnZlY3RvcnRpbGVzIiwiYSI6ImNpZnlkb2d0dTAyaDd1M2x6ZGg5cGp4aXQifQ.74RX5npQKmCE1jxsbXYNzQ|47.35929250871075|8.562469482421875|13

Mapbox Streets Tile Inspector:

http://klokantech.github.io/ol3-sandbox/vector/tile-inspector.html#http://a.tiles.mapbox.com/v4/mapbox.mapbox-streets-v5.json?access_token=pk.eyJ1Ijoib3NtMnZlY3RvcnRpbGVzIiwiYSI6ImNpZnlkb2d0dTAyaDd1M2x6ZGg5cGp4aXQifQ.74RX5npQKmCE1jxsbXYNzQ|47.389696670088995|368.5221578025774|13

**Comment by manuelroth** [comment]
@petrsloup Thanks this helps a lot. I like that we can see the tileinfo with all the classes and types on the right side.

**Comment by klokan** [comment]
The TileJSON for V5 and V6 are mentioned at: https://github.com/geometalab/osm2vectortiles/issues/25#issuecomment-151569602

This viewer had to be changed from JSONP to XHR/AJAX.
We have added also permalink lat/lon/zoomlevel functionality for easier linking to a certain place - if we discuss in future one particular tile.

Closing this ticket. Your improvements and contributions to Debug Viewer are welcome via pull request.
